src/integrators.jl

```diff
@@ -1,152 +1,300 @@
-            adjust_dt!(integrator, integrator.tstop - integrator.t)
-        end
-        DiffEqBase.step!(integrator)
-
-        if integrator.step == integrator.stepstop
-            break
-        end
-    end
-
-    if isdefined(DiffEqBase, :finalize!)
-        DiffEqBase.finalize!(integrator.callback, integrator.u, integrator.t, integrator)
-    end
-    if isempty(integrator.sol.t)
-        push!(integrator.sol.t, integrator.t)
-        push!(integrator.sol.u, integrator.u)
+    while !isempty(integrator.tstops) && integrator.step != integrator.stepstop
+        __step!(integrator)
     end
+    DiffEqBase.finalize!(
+        integrator.callback,
+        integrator.u,
+        integrator.t,
+        integrator,
+    )
     return integrator.sol
 end
 
-
 # either called directly, or via solve!
 function DiffEqBase.step!(integrator::DistributedODEIntegrator)
-    step_u!(integrator) # solvers need to define this interface
-    integrator.t += integrator.dt
+    if integrator.advance_to_tstop
+        tstop = first(integrator.tstops)
+        while !reached_tstop(integrator, tstop)
+            __step!(integrator)
+        end
+    else
+        __step!(integrator)
+    end
+end
+
+# helper functions for dealing with time-reversed integrators in the same way
+# that OrdinaryDiffEq.jl does
+tdir(integrator) =
+    integrator.tstops.ordering isa DataStructures.FasterForward ? 1 : -1
+is_past_t(integrator, t) =
+    tdir(integrator) * (t - integrator.t) < zero(integrator.t)
+reached_tstop(integrator, tstop) = integrator.t == tstop ||
+    (!integrator.dtchangeable && is_past_t(integrator, tstop))
+
+function __step!(integrator)
+    tstops = integrator.tstops
+    tstop = first(tstops)
+
+    # update step and dt before incrementing u; if dt is changeable and there is
+    # a tstop within dt, reduce dt to tstop - t
     integrator.step += 1
+    integrator.dt = integrator.dtchangeable ?
+        tdir(integrator) * min(integrator._dt, abs(tstop - integrator.t)) :
+        tdir(integrator) * integrator._dt
+    step_u!(integrator)
+
+    # increment t by dt, rounding to the nearest tstop if that is roughly
+    # equivalent up to machine precision; the specific bound of 100 * eps...
+    # is taken from OrdinaryDiffEq.jl
+    t_plus_dt = integrator.t + integrator.dt
+    t_unit = oneunit(integrator.t)
+    max_t_error = 100 * eps(float(integrator.t / t_unit)) * t_unit
+    integrator.t = abs(tstop - t_plus_dt) < max_t_error ? tstop : t_plus_dt
 
     # apply callbacks
     discrete_callbacks = integrator.callback.discrete_callbacks
     for callback in discrete_callbacks
-        if callback.condition(integrator.u,integrator.t,integrator)
+        if callback.condition(integrator.u, integrator.t, integrator)
             callback.affect!(integrator)
         end
     end
+
+    # remove tstops that were just reached
+    while !isempty(tstops) && reached_tstop(integrator, first(tstops))
+        pop!(tstops)
```

Comment:
This is an allocating function.

Reply:
Can we re-design this to be non-allocating?

Reply:
tstops can be dynamically modified as the integrator runs, and I'm not sure how to do that without push!/pop!. I could have used an array, but OrdinaryDiffEq uses a heap, so I just copied them.

Reply:
Ok, I guess we can merge for now. I'll see if this can be addressed later.